Add byte_size and kind_byte_size to Bigarray

objmagic opened the pull request:
I removed my last commit and made this patch simpler
- This patch **adds** `byte_size` and `kind_byte_size` APIs to `Bigarray` module.
  - `kind_byte_size` is implemented by hardcoding byte size information. 
  - `byte_size` is implemented by using function `kind_byte_size`
- This patch **resolves** [Mantis 6263](http://caml.inria.fr/mantis/view.php?id=6263)
- Test cases can be found under `/testsuite/tests/lib-bigarray`

objmagic committed:
Add byte_size and kind_byte_size to Bigarray

objmagic commented:
ping?

damiendoligez commented:
I'm OK for merging this, but why do you say that `byte_size` is implemented by wrapping `caml_ba_byte_size`? As far as I can tell from your patch, you're simply recomputing the size in OCaml.

objmagic commented:
@damiendoligez sorry I made a mistake in PR message. They are both implemented as normal OCaml functions.

damiendoligez commented:
I'd like to get @xavierleroy's opinion on this PR.

Also, shouldn't `kind_byte_size` be called `byte_size_of_kind`?

dbuenzli commented:
> Also, shouldn't kind_byte_size be called byte_size_of_kind?

I would say no for the following reasons:
1. `byte_size` is not a type, in this case `byte_size` is a (derived) property of the type `Bigarray.kind` we are not really "converting" a value of type `kind` to a value of type `byte_size`. If that what in is its own module I'm sure you would have written `Kind.byte_size` and not `Kind.to_byte_size`.
2. If you really want that then `Bigarray.byte_size` should be called `Bigarray.to_byte_size` which I personally find odd.

What I feel could be disputed though is the term `size` which could be replaced by `length`. Witnesses `String.length`, `Array.length`, `Stack.length`, `Bytes.length`, `Buffer.length`, `Queue.length`, `Hashtbl.length`. I personally use `byte_length` in the context of bigarrays in at least one API [here](http://erratique.ch/software/gg/doc/Gg.Ba.Buffer.html#VALbyte_length). My proposal in mantis was certainly influenced by the C function which uses `size` in the name but I think `length` would be better.

damiendoligez commented:
> What I feel could be disputed though is the term `size` which could be replaced by `length`.

Either that, or we decide that `length` is in natural units while `size` is always in bytes, and we just use `size` instead of `byte_size`.

dbuenzli commented:
> we just use size instead of byte_size.

I really prefer explicit units, it makes code much more straightforward and unambiguous when you read it. No convention to remember. Also the problem is that foreign code may easily overlook such a subtle convention and use `size` for non-byte sizes which will make things overall ambiguous even if your own library managed to be consistent in its usage. So rather `byte_length` for me.

damiendoligez commented:
OK for `byte_length`.

xavierleroy commented:
I'm OK with this addition to Bigarray.  The original [PR#6263](http://caml.inria.fr/mantis/view.php?id=6263) could have given an actual use case instead of a vague "is useful in various contexts".  

Concerning the name controversy, I tend to think of the length of an array as the number of elements in the array, and of the size of an array as the amout of storage used.  So, if it is still time to voice an opinion, "size", or if you must "byte size", sounds better to me than "byte length".

objmagic commented:
stopped renaming halfway... I'll wait when everyone reaches consensus

objmagic commented:
and I thought Bigarray is not a part of stdlib yet? why label it as stdlib?

damiendoligez commented:
> and I thought Bigarray is not a part of stdlib yet? why label it as stdlib?

We are seriously considering moving `bigarray` into `stdlib`, and the byte/length naming discussion is definitely related to the stdlib.

It doesn't mean this PR stay on hold indefinitely, it just means we need to be careful in the context of the stdlib discussions.

dbuenzli commented:
> I tend to think of the length of an array as the number of elements in the array, and of the size of an array as the amout of storage used. 

I don't know if there's an english subtlety or a highly widespread convention that I'm missing here, but for me both are exactly interchangeable and I never saw any consistent usage in the wild of these terms; along with those that use also `count`.

> So, if it is still time to voice an opinion, "size", or if you must "byte size"

Specifying units avoids a trip to the documentation or your memory when you read. Especially since we already have `Obj.size` whose units are not bytes.

objmagic commented:
by the way, I feel like `length` applies best on one-dimensional thing. Should we also use this word on two-dimensional or three-dimensional arrays?

dbuenzli commented:
> by the way, I feel like length applies best on one-dimensional thing. Should we also use this word on two-dimensional or three-dimensional arrays?

I think that's a good argument in favor of `byte_size`. Regarding the demanded use cases here are a few occurences:
- [in `Gg`](https://github.com/dbuenzli/gg/blob/455caf7318921b98ef384827fef9117be0fdbddd/src/gg.ml#L2911-L2925)
- [in `Tgls`](https://github.com/dbuenzli/tgls/blob/9c2e8ac42924521c945824165ced10c5cf308710/support/gen.ml#L230-L243)
- [in `Tsdl`](https://github.com/dbuenzli/tsdl/blob/7de867c01575e2e2bd67914cc8cd5c8ad3cf322a/src/tsdl.ml#L138-L148)

damiendoligez commented:
OK for `byte_size` then.

Commit:
Update changelog

mshinwell commented:
I think we should use the term "size_in_bytes" rather than "byte_size".  The phrase "byte size" sounds like something to do with the size of an individual byte.

Also, I suggest adding some more detail to the comment in the .mli.  What exactly does the "size in bytes" measure?  It should say whether it includes the overhead of the enclosing OCaml value, for example.

gasche commented:
We discussed this at the developer meeting yesterday, and agree to merge it as soon as a name that pleases native english speakers is found.

objmagic commented:
It does not count header and only counts the size of array data in bytes. I'll add more comments and update any term usage when everyone reaches some consensus on term usage.

dbuenzli commented:
I trust @mshinwell

objmagic commented:
@dbuenzli is `[size_in_bytes a] is size of [a]'s data in bytes, excluding [a]'s headers.` fine? I want to make sure my comment is very clear.

dbuenzli commented:
> @dbuenzli is [size_in_bytes a] is size of [a]'s data in bytes, excluding [a]'s headers. fine? I want to make sure my comment is very clear.

The notion of header is not mentioned anywhere in the docs so it's likely to be confusing. Here's my take on it: 

```
[kind_size_in_bytes k] is the number of bytes used to store an element of type [k].
[size_in_bytes a] is the number of elements in [a] multiplied by [a]'s {!kind_size_in_bytes}.
```

Commit:
Change function names, improve documentation.

Change ``byte_size'' to ``size_in_bytes``, change ``kind_byte_size``
to ``kind_size_in_bytes``.

Add detailed comment indicating size functions do not consider header
of OCaml value.

objmagic commented:
@dbuenzli Thank you very much, Daniel. I have updated my patch.

@mshinwell @gasche I think this patch is done. Please make a final review. :)

dbuenzli commented on Changes:
```diff
@@ -101,6 +101,8 @@ Other libraries:
 - PR#6896: serious reimplementation of Big_int.float_of_big_int and
   Ratio.float_of_ratio, ensuring that the result is correctly rounded.
   (Xavier Leroy)
+- PR#6263: add kind_byte_size and byte_size functions to Bigarray module.
```
You forgot to update this.

Commit:
Update Changelog

Commit:
Add Shinwell as reviewer

damiendoligez commented on testsuite/tests/lib-bigarray/bigarrays.ml:
```diff
@@ -410,6 +410,12 @@ let _ =
 
+  testing_function "size_in_bytes_one";
+  test 1 (Array1.size_in_bytes (from_list int [1;2;3;4;5])) 40;
+  test 2 (Array1.size_in_bytes (from_list int [])) 0;
```
The two lines above look like they are going to fail on a 32-bit architecture.

Likewise for three more tests a bit further below.

Commit:
Fix test

Several tests will fail on 32-bit machine.

damiendoligez commented:
Merged (commit 192ccd090239ad20160931e79e2adcea847e6f90).

Thanks for the report and patch.
